internal/ot/naor_pinkas.go

```diff
@@ -0,0 +1,169 @@
+	secretR, pointR, err := crypto.GenerateKey()
+	if err != nil {
+		return fmt.Errorf("error generating keys: %w", err)
+	}
+
+	// send point A to receiver
+	if err := writer.Write(pointA); err != nil {
+		return fmt.Errorf("error writing point: %w", err)
+	}
+
+	// send point R to receiver
+	if err := writer.Write(pointR); err != nil {
+		return fmt.Errorf("error writing point: %w", err)
+	}
+
+	// precompute A = rA
+	pointA = pointA.ScalarMult(secretR)
+
+	// encrypt plaintext messages and send them.
+	for i := range otMessages {
+		// receive key material
+		keyMaterial := crypto.NewPoint()
+		if err := reader.Read(keyMaterial); err != nil {
+			return fmt.Errorf("error reading point: %w", err)
+		}
+		var keys [2]*crypto.Point
+		// compute and derive key for first OT message
+		// K0 = rK0
+		keys[0] = keyMaterial.ScalarMult(secretR)
+		// compute and derive key for second OT message
+		// K1 = rA - rK0
+		keys[1] = pointA.Sub(keys[0])
+
+		// encrypt plaintext message with keys
+		for choice, plaintext := range otMessages[i] {
+			// encryption
+			ciphertext, err := crypto.XorCipherWithBlake3(keys[choice].DeriveKeyFromECPoint(), uint8(choice), plaintext)
+			if err != nil {
+				return fmt.Errorf("error encrypting sender message: %w", err)
+			}
+
+			// send ciphertext
+			if _, err = rw.Write(ciphertext); err != nil {
+				return fmt.Errorf("error writing bytes: %w", err)
+			}
+		}
+	}
+
+	return
+}
+
+func (n naorPinkas) Receive(choices []uint8, messages [][]byte, rw io.ReadWriter) (err error) {
+	if len(choices)*8 != len(messages) || len(choices)*8 != len(n.msgLen) {
+		return ErrBaseCountMissMatch
+	}
+
+	// instantiate Reader, Writer
+	reader := crypto.NewECPointReader(rw)
+	writer := crypto.NewECPointWriter(rw)
+
+	// receive point A from sender
+	pointA := crypto.NewPoint()
+	if err := reader.Read(pointA); err != nil {
+		return fmt.Errorf("error reading point: %w", err)
+	}
+	// recieve point R from sender
+	pointR := crypto.NewPoint()
+	if err := reader.Read(pointR); err != nil {
+		return fmt.Errorf("error reading point: %w", err)
+	}
+
+	// Generate points B, 1 for each OT
+	bSecrets := make([][]byte, len(messages))
+	for i := range messages {
+		var pointB = crypto.NewPoint()
+		// generate receiver priv/pub key pairs going to take a long time.
+		bSecrets[i], pointB, err = crypto.GenerateKey()
+		if err != nil {
+			return fmt.Errorf("error generating keys: %w", err)
+		}
+
+		// for each choice bit, compute the key material corresponding to
+		// the choice bit and sent it.
+		if !util.IsBitSet(choices, i) {
+			// K0 = Kc = B
+			if err := writer.Write(pointB); err != nil {
+				return fmt.Errorf("error writing point: %w", err)
+			}
+		} else {
+			// K1 = Kc = B
+			// K0 = K1-c = A - B
+			if err := writer.Write(pointA.Sub(pointB)); err != nil {
+				return fmt.Errorf("error writing point: %w", err)
+			}
+		}
+	}
+
+	// receive encrypted messages, and decrypt it.
+	for i := range messages {
```

Comment:
This implementation of naor-pinkas will deadlock with a larger baseCount or with a io.ReadWriter that doesn't buffer behind the scene. The problem is that the sender writes while the read doesn't read until it is done writing. One can expose the bug by increasing base count to 1 << 15 and running the test for naor-pinkas. 1 << 14 takes a few seconds to complete, 1 << 15 blocks forever. The solution to expose these bugs is to use io.ReadWriter that don't buffer such as wrapping a pair of io.Pipe.